File: leetcode/56. Merge Intervals.cpp
// Author: Shengjia Yan
// Date: 2018-05-11 Friday
// Email: i@yanshengjia.com
// Time Complexity: O(n)
// Space Complexity: O(n)

/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    static bool my_compare(const Interval &a, const Interval &b)
    {
        return a.start < b.start;
    }
    
    vector<Interval> merge(vector<Interval>& intervals) {
        int length = intervals.size();
        vector<Interval> res;
        
        sort(intervals.begin(), intervals.end(), my_compare);
        
        for(int i=0; i<length; i++) {
            if (res.empty())
                res.push_back(intervals[i]);
            else {
                int size = res.size();
                if (res[size-1].start <= intervals[i].start && intervals[i].start <= res[size-1].end) {
                    int max = intervals[i].end > res[size-1].end ? intervals[i].end : res[size-1].end;
                    res[size-1].end = max;
                }
                else {
                    res.push_back(intervals[i]);
                }
            }
        }
        return res;
    }
};

File: leetcode/Sort/56. Merge Intervals.py
from typing import List

# Sort intervals by their start number, then merge overlapping intervals
# Time Complexity: O(NlogN), where N is the number of intervals. Sorting takes O(NlogN) time.
# Space Complexity: O(1), not counting the space used by the output list.
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        
        intervals.sort(key=lambda x: x[0]) # list.sort() returns None, sorting in place

        res = []

        for interval in intervals:
            # res is empty or last interval in res is NOT overlapping with current interval
            # append interval into res
            if len(res) == 0 or res[-1][1] < interval[0]:
                res.append(interval)
            else:
                # last interval in res is overlapping with current internval
                # merge them
                res[-1][1] = max(res[-1][1], interval[1])
        return res